### ISSUE TITLE
1.8.0 requires code changes around functions taking git_commit

### ISSUE THREAD

**Issue by carrotIndustries** [libgit2/libgit2#6793]
### Reproduction steps

Compile code that was written against version prior to 1.8.0 such as 

https://github.com/horizon-eda/horizon/blob/5300ccd7f9ba81818eb34d9ffc1b8fe80df6c658/src/pool-prj-mgr/pool-mgr/pool_git_box.cpp#L634-L642

```c++
std::array<const git_commit *, 2> parents;
parents.at(0) = master_commit;
parents.at(1) = pr_commit;
if (git_commit_create(&new_commit_oid, repo, "HEAD", signature, signature, NULL,
                      ("merge pr " + std::to_string(pr_id)).c_str(), tree, parents.size(), parents.data())
    != 0) {
    throw std::runtime_error("error committing");
}
git_repository_state_cleanup(repo);
```

### Expected behavior

Should compile, the release notes don't include any related breaking changes in that regard.

### Actual behavior

Doesn't compile, see https://github.com/horizon-eda/horizon/actions/runs/8589355641/job/23582932883#step:6:559

The offending commit is https://github.com/libgit2/libgit2/commit/cf19ddc52227f4ec2efd4c0a84aa5d2362f0ffc7

For the `git_commit_create` method, I'm not quite sure if this commit actually did the right thing:

 - As far as I can tell, `git_commit_create` should not have to modify the passed `git_commit` objects, so the `const` was in fact correct
 - I think `const git_commit * const parents[]` is what we need. This also doesn't have the compatibility problem 

The current situation would require some annoying version check macros and casts to be compatible with versions prior to 1.8.0 and 1.8.0.

### Version of libgit2 (release number or SHA1)

1.8.0

### Operating system(s) tested

Linux

**Comment by ethomson** [comment]
Right, I wasn't intending or expecting this change to be quite so troublesome. Indeed, I can imagine that C++ and const correctness would require some changes. I didn't realize that C++ enforced const correctness on `extern C` but that's little surprise as I'm not a C++ programmer and this is not a C++ library.

In any case, `git_commit`s _aren't const_. They have an internal state that points into an internal cache, those pointers can change, and they're refcounted which _will_ change.

However:
1. There are a surprising number of places that take a `const git_commit *`. Which surprises me.
2. I wasn't intending for this to be a breaking change. (Even if you are using C++.)

I need to have a deeper look at this, but I wanted to provide the rationale behind the change.

**Comment by carrotIndustries** [comment]
For reference, I've added a workaround in the meantime: https://github.com/horizon-eda/horizon/commit/af6c626b690aa579966f67056d8cf1fdc143a417

**Comment by ethomson** [comment]
Fixed via #6829

**Comment by StefanBruens** [comment]
> Right, I wasn't intending or expecting this change to be quite so troublesome. Indeed, I can imagine that C++ and const correctness would require some changes. I didn't realize that C++ enforced const correctness on `extern C` but that's little surprise as I'm not a C++ programmer and this is not a C++ library.
> 
> In any case, `git_commit`s _aren't const_. They have an internal state that points into an internal cache, those pointers can change, and they're refcounted which _will_ change. Making this `const git_commit **` was a mistake in the first place, which you can see by the unnecessary casts that existed in the examples, etc.

If the *observable* behavior does not depend on the exact internal state of the cache and on the refcounts, then git_commit can and should be `const`.

A good read (including a mutable cache for an otherwise const object) can be found here:
https://herbsutter.com/2013/05/28/gotw-6b-solution-const-correctness-part-2/
